[Diamondroxxx] iP

EepyKitty opened the pull request:
## CatBot

> This is a fancy quote that exists for no reason 

CatBot has a few key properties
* it's **fast**
* it's **efficient**
* it's a **cat**!
* ~~it also does your homework~~ _(unfortunately this is a joke)_

To use it, follow these easy steps:

1. Download the latest release from [here](https://github.com/Diamondroxxx/ip/releases/tag/A-Jar).
2. Navigate to the folder using a terminal
3. Run `java -jar catbot.jar`
4. Enjoy! 🎉

Current commands include

- [x] todo
- [x] deadline
- [x] event
- [x] list
- [x] mark
- [ ] unmark
- [x] echo
- [ ] find

-----
### For Java programmers

If you happen to be a Java programmer and want to integrate `CatBot` into your program, here is how to make a `CatBot` instance.

```java
public static void main(String[] args) {
    CatBot bot = new CatBot("save/file/location");
}
```

Commit:
Add Gradle support

Commit:
Finish Level 1

Commit:
Finish Level 6

Commit:
Add Echo Command.

Commit:
Fix Bug With Box Width Including Coloured Text

Commit:
Add Tests

Commit:
Add Ability to Save to Disk

Commit:
Fix Style a bit

Commit:
Add LocalDateTime Support

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge Branches

Commit:
Refactor code into packages

Commit:
Fix package naming

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Add Gradle support for building to a jar file

Commit:
Add junit tests

Commit:
Add find command

Commit:
Merge branch 'branch-Level-9'

swxk19 commented on src/main/java/catbot/CatBotException.java:
```diff
@@ -0,0 +1,15 @@
+    public CatBotException(String message) {
+        super(message);
+    }
+}
```
CatBotException can be split into more specific exceptions, so a specific exception can be thrown instead of general error. e.g: ListIndexOutOfRangeException if the user tries to invoke a command outside of the list's range

swxk19 commented on src/main/java/catbot/parser/Parser.java:
```diff
@@ -0,0 +1,112 @@
+                return new AddCommand(cmd[1].strip());
+            } catch (ArrayIndexOutOfBoundsException e) {
+                throw new CatBotException("That's the wrong format!");
+            }
```
It might be better to move the error handling to within the individual method commands
So that in future, if you decide to do it, you can have more specific error handling for each kind of task created. E.g: "That's the wrong format for an Event! Event should have the format: event <task_name> /from ....."

swxk19 commented on src/main/java/catbot/tasklist/EventTask.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString()
+                + " (" + formatDate(startsAt)
+                + " – " + formatDate(endsAt) + ")";
```
Might be cleaner to use String.format() :)

swxk19 reviewed (commented):
Great Job!

Bisceto commented on src/main/java/catbot/commands/Command.java:
```diff
@@ -0,0 +1,28 @@
+
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws CatBotException;
+
+    public void loadCommand(ArrayList<Task> tasks) throws CatBotException {}
```
I think you can declare loadCommand as an abstract method.

Bisceto commented on src/main/java/catbot/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,36 @@
+ * Handles deleting a task from the task list.
+ */
+public class DeleteCommand extends Command {
+    private final int index;
```
Since this attribute is final, it may be better to make 'index' all uppercase

Bisceto commented on src/main/java/catbot/parser/Parser.java:
```diff
@@ -0,0 +1,112 @@
+     * @throws CatBotException if the user input is malformed.
+     */
+    public static Command parse(String command) throws CatBotException {
+        String[] cmd = command.split(" ", 2);
```
Variable names for collections should be in plural (and meaningful as well)

Bisceto commented on src/main/java/catbot/parser/Parser.java:
```diff
@@ -0,0 +1,112 @@
+        default:
+            throw new CatBotException("I don't know what you mean.");
+        }
+    }
```
Good use of switch statements in parsing the command! If you can handle the exceptions internally with each command, it will definitely make the code look cleaner

Bisceto commented on src/main/java/catbot/storage/Storage.java:
```diff
@@ -0,0 +1,92 @@
+
+            scanner.close();
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
```
You can throw a CatBotException to make the error handling more meaningful.

Bisceto commented on src/main/java/catbot/tasklist/Task.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public abstract class Task {
+    protected boolean isDone;
+    final String description;
```
You should make 'description' either protected or private. By default, Java leaves unspecified variables as public and this can violate information hiding

Bisceto reviewed (commented):
LGTM! Just a few nits to fix.

Commit:
Make UI

Commit:
Delete files only needed for CLI UI

Commit:
Add assertions

CatBot lacks assertions to ensure code reliability. Without assertions,
catching potential bugs and errors is time-consuming and error-prone.

Let's add assertions to verify that key assumptions are being met
throughout the execution of the program.

This makes errors show up early and loudly so that it is easier to spot
the source of the issue.

Commit:
Split AddCommand into three separate classes for better code style consistency

The AddCommand class handles adding various types of tasks, violating
the principle of consistent levels of abstraction.

This makes it harder to maintain and read the code.

Let's implement three new classes, AddTodoCommand, AddDeadlineCommand,
and AddEventCommand, each with a single responsibility for adding their
respective task types.

This approach breaks down complex tasks into smaller, more manageable
pieces with a single responsibility for easier understanding and
maintenance over time.

Commit:
Run SaveCommand on close by pressing x button

Commit:
Fix small error in comment

EepyKitty committed:
Merge pull request #2 from Diamondroxxx/branch-A-Assertions

Add assertions

Commit:
Fix small error in comment

Commit:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/catbot/parser/Parser.java

EepyKitty committed:
Merge pull request #3 from Diamondroxxx/branch-A-CodeQuality

Improve Code Quality

Commit:
Add support for recurring tasks

EepyKitty committed:
Merge pull request #4 from Diamondroxxx/branch-B-RecurringTasks

Add support for recurring tasks

Commit:
Fix bug with truncating list

Commit:
Remove bye command

The bye command wasn't actually terminating the program.

Let's just get rid of it as closing via the x button is more intuitive.

Commit:
Factor out formatDate into Task

Commit:
Make window resizable

Commit:
Fix user image to be cropped correctly

EepyKitty committed:
Merge pull request #5 from Diamondroxxx/branch-A-BetterGui

Make the Gui better

Commit:
Add Ui.png

Commit:
Add natural language date parsing

EepyKitty committed:
Merge pull request #6 from Diamondroxxx/branch-BetterDateParsing

Add Natural Language Date Parsing

Commit:
Add User Guide

Commit:
Add User Guide

EepyKitty committed:
Create _config.yml
